### PR TITLE
refactor plugin: es6 + comments + fit documentation

### PR DESCRIPTION
**Pull request by stafyniaksacha**
## Description
- Refactor plugin with ES6 Syntax
- Update `docker-compose` with latest `elasticsearch` image
- Fit samples with `TL;DR; examples plugins` from documentation
- Added a lot of links to the new documentation
- Added `manifest.json` file

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -1,123 +1,239 @@
...
+          // The options provided to the authenticate function during the authentication process
+          authenticateOptions: {
+            scope: []
+          },
+
+          // The list of fields that may be provided in the credentials
+          fields: ['login', 'password']
+        },
+        methods: {
+          // The name of the create function
```
Comment (scottinet) [inline lib/index.js]: Erroneous function description.

Also, I would remove all the `[mandatory]` indicator on other functions descriptions, and instead add a `(optional)` one in this one

Comment (scottinet) [inline lib/index.js]: Moreover, these descriptions aren't helpful. I would use a short description (copy-paste from the documentation works here), and a link to the official documentation

File: lib/index.js [hunk 2]
```diff
@@ -1,123 +1,239 @@
-// Used to show how to extend the authentication strategies,
-// you probably want to change or drop this.
-// See also http://docs.kuzzle.io/plugin-reference/#authentication-plugin
-const LocalStrategy = require('passport-local').Strategy
+/**
+ * Plugins must be valid NodeJS require-able modules, usually shipped as a directory containing either:
```
Comment (ballinette) [inline lib/index.js]: plugins are not node-modules anymore

Comment (scottinet) [inline lib/index.js]: That's not exactly right.
Plugins root directory must contain either a valid package.json file or a index.js one.

File: lib/index.js [hunk 3]
```diff
@@ -1,123 +1,239 @@
...
+   * @param {Request} request The request that triggered the event
    * @param {Function} callback The callback that bears the result of the
    *                            function. Signature: `callback(error, request)`
-   * @see http://docs.kuzzle.io/plugin-reference/#pipe-plugins
+   *
+   * @see http://docs.kuzzle.io/plugins-reference/plugins-features/adding-pipes/
+   * @see http://docs.kuzzle.io/guide/essentials/request-and-response-format/
   */
-  this.syncListener = function (request, callback) {
+  syncListener (request, callback) {
```
Comment (ballinette) [inline lib/index.js]: nitpicking: should be intersting to add also another sync listener sample with another parameter type than `request`.
For example, adding a pipe to a `notify` event (see https://github.com/kuzzleio/documentation/blob/rc.x/src/kuzzle-events/notify/index.md )

File: lib/index.js [hunk 4]
```diff
@@ -1,123 +1,269 @@
...
-    'document:beforeCreateOrReplace': 'asyncListener',
-    'document:beforeReplace': 'asyncListener',
-    'document:beforeUpdate': 'asyncListener',
-  }
-
-  /**
-   * Specifies a set of events along with the synchronous
-   * listener function they trigger.
+   * Workflow is:
+   *  - Kuzzle load plugins in <kuzzle install dir>/plugins/enabled/* and try to instanciate all ones found
```
Comment (xbill82) [inline lib/index.js]: "load" :point_right: "loads"
"try" :point_right: "tries"
"all ones found in this step" :point_right: "them"

File: lib/index.js [hunk 5]
```diff
@@ -1,123 +1,269 @@
...
+   *    during this step, plugin is only loaded in kuzzle memory (constructor is called)
```
Comment (xbill82) [inline lib/index.js]: don't get the meaning of "plugin is only loaded in kuzzle memory (constructor is called)"

File: lib/index.js [hunk 6]
```diff
@@ -1,123 +1,269 @@
...
+   *  - Kuzzle try to initialiate worker plugins (this.config.threads = Number > 0) in spawned process thought pm2
```
Comment (xbill82) [inline lib/index.js]: "try"  :point_right: "tries"
"thought"  :point_right: "through"

File: lib/index.js [hunk 7]
```diff
@@ -1,123 +1,269 @@
...
+   *  - "init" function is called for both woker and regular plugins.
+   *  - Plugin manager register all plugin features into kuzzle (register hooks/pipes, load custom configuration, check for authentication strategies, ...)
```
Comment (xbill82) [inline lib/index.js]: "register" :point_right: "registers"

File: lib/index.js [hunk 8]
```diff
@@ -1,123 +1,269 @@
...
+     * which you can merge with overrided configuration received by Kuzzle during plugin initialization
+     *
+     * @type {Object}
+     */
+    this.config = {
+      param: '<default value>'
+    };
+
+    /**
+     * Define if this plugin support worker plugins (multi-threaded)
```
Comment (xbill82) [inline lib/index.js]: "support" :point_right: "supports"

File: lib/index.js [hunk 9]
```diff
@@ -1,123 +1,269 @@
...
+     * Setting this to true allow users to set "threads" configuration
```
Comment (xbill82) [inline lib/index.js]: "allow" :point_right: "allows"

File: lib/index.js [hunk 10]
```diff
@@ -1,123 +1,269 @@
...
+     *
+     * /!\ warning: allowing threads result in incapacity to inject:
```
Comment (xbill82) [inline lib/index.js]: I would reformulate this as
"Threaded plugins will not be able to access:"

File: lib/index.js [hunk 11]
```diff
@@ -1,123 +1,269 @@
...
+     * @see  http://docs.kuzzle.io/plugins-reference/plugins-features/adding-pipes/
+     * @see http://docs.kuzzle.io/kuzzle-events/
+     */
+    this.pipes = {
+      'document:beforeCreate': 'syncListener',
+      'realtime:beforePublish': 'syncListener'
+    };
+
+    /**
+     * This exposed "controllers" property tells Kuzzle that it needs to extend
```
Comment (xbill82) [inline lib/index.js]: I would reformulate as
"This exposed "controllers" property enables to extend the Kuzzle API with new controllers and actions"

File: lib/index.js [hunk 12]
```diff
@@ -1,123 +1,269 @@
...
+    this.controllers = {
+      'myNewController': {
+        'myNewAction': 'saySomething'
+      }
+    };
+
+    /**
+     * We also want to expose our new controller to the HTTP protocol.
+     * To do so, we give Kuzzle instructions on how we want to expose our
+     * controller to HTTP.
```
Comment (xbill82) [inline lib/index.js]: I would rather say
"The following "routes" exposed property allows Kuzzle to bind new controllers and actions to HTTP endpoints"

**Pull request merged**